presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java

```diff
@@ -0,0 +1,219 @@
+        implements SpillCipher
+{
+    public enum Mode
+    {
+        AES_128(128),
+        AES_256(256);
+
+        private final int keyBits;
+
+        Mode(int keyBits)
+        {
+            this.keyBits = keyBits;
+        }
+
+        public int getKeyBits()
+        {
+            return keyBits;
+        }
+    }
+
+    private SecretKeySpec key;
+    private final int keyBits;
+    private final int ivBytes;
+
+    AESSpillCipher(Mode mode)
+    {
+        this.keyBits = requireNonNull(mode, "mode is null").getKeyBits();
+        this.key = generateNewSecretKey(keyBits);
+        this.ivBytes = createEncryptCipher().getIV().length;
+    }
+
+    @Override
+    public ByteBuffer decrypt(ByteBuffer cipherText)
+    {
+        int length = cipherText.remaining();
+        if (length < ivBytes) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, format("cipherText too small to contain IV. Minimum: %s, Actual: %s", ivBytes, length));
+        }
+        byte[] iv = new byte[ivBytes];
+        cipherText.get(iv);
+        Cipher cipher = createDecryptCipher(new IvParameterSpec(iv));
+        ByteBuffer output = ByteBuffer.allocate(cipher.getOutputSize(cipherText.remaining()));
+        try {
+            cipher.doFinal(cipherText, output);
+            output.flip();
+            return output;
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failure during decrypt", e);
+        }
+    }
+
+    @Override
+    public byte[] decrypt(byte[] cipherText)
+    {
+        return unwrapOrCopyToByteArray(decrypt(ByteBuffer.wrap(cipherText)));
+    }
+
+    @Override
+    public ByteBuffer encrypt(ByteBuffer plainText)
+    {
+        Cipher cipher = createEncryptCipher();
+        byte[] iv = cipher.getIV();
+        ByteBuffer output = ByteBuffer
+                .allocate(iv.length + cipher.getOutputSize(plainText.remaining()))
+                .put(iv);
+        try {
+            cipher.doFinal(plainText, output);
+            output.flip();
+            return output;
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failure during encrypt", e);
+        }
+    }
+
+    @Override
+    public byte[] encrypt(byte[] plainText)
+    {
+        return unwrapOrCopyToByteArray(encrypt(ByteBuffer.wrap(plainText)));
+    }
+
+    @Override
+    public void destroy()
+    {
+        this.key = null;
+    }
+
+    @Override
+    public boolean isDestroyed()
+    {
+        SecretKeySpec key = this.key;
+        return key == null || key.isDestroyed();
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("keyBits", this.keyBits)
+                .add("isDestroyed", isDestroyed())
+                .toString();
+    }
+
+    private SecretKeySpec verifyKeyNotDestroyed()
+    {
+        SecretKeySpec key = this.key;
```

Comment:
This variable is not needed as well as the `else` below:

```
        if (key == null || key.isDestroyed()) {
            throw new PrestoException(GENERIC_INTERNAL_ERROR, "SpillCipher already destroyed");
        }
        return key;
```

Also, this method could just return void and the callers could simply use this.key:

```
    private void verifyKeyNotDestroyed()
    {
        if (key == null || key.isDestroyed()) {
            throw new PrestoException(GENERIC_INTERNAL_ERROR, "SpillCipher already destroyed");
        }
    }

    private Cipher createEncryptCipher()
    {
        verifyKeyNotDestroyed();
        Cipher cipher = createUninitializedCipher();
        try {
            cipher.init(Cipher.ENCRYPT_MODE, key);
            return cipher;
        }
        catch (GeneralSecurityException e) {
            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failed to initialize cipher for encryption", e);
        }
    }
```

Reply:
See [my explanation above](https://github.com/prestodb/presto/pull/12617#discussion_r277841598) for the rationale behind using a local variable and returning it from `verifyKeyNotDestroyed()`.